File: codegen/codegen-erlang/src/test/resources/ir/endpoint_rules_module.expected.erl
%% Generated endpoint rule resolver for smithy.beam.test.endpoints#EndpointRulesService.
-module(endpoint_rules_service_endpoints).
-include("runtime_types.hrl").
-export([resolve/2]).
-type client_config() :: #{binary() => term()}.
-type endpoint_params() :: #{binary() => term()}.

-spec resolve(client_config(), endpoint_params()) -> {ok, #{url := binary()}} | {error, term()}.
resolve(Config, Params) ->
    aws_endpoint_rules:evaluate(?ENDPOINT_RULE_SET, merge_params(Config, Params)).

merge_params(Config, Params) ->
    ConfigParams = config_to_rule_params(Config),
    ClientParams = client_context_params(Config),
    maps:merge(maps:merge(ConfigParams, ClientParams), Params).

config_to_rule_params(Config) ->
    case maps:get(region, Config, undefined) of
        undefined -> #{};
        Value -> #{<<"Region">> => Value}
    end.

client_context_params(Config) ->
    maps:merge(
        optional_param(Config, region, <<"Region">>), optional_param(Config, bucket, <<"Bucket">>)
    ).

optional_param(Config, Key, RuleKey) ->
    case maps:get(Key, Config, undefined) of
        undefined -> #{};
        Value -> #{RuleKey => Value}
    end.

File: codegen/codegen-elixir/src/test/resources/ir/endpoint_rules_module.expected.ex
defmodule EndpointRulesServiceEndpoints do
  @moduledoc "Generated endpoint rule resolver for Smithy service clients."
  alias RuntimeTypes, as: RuntimeTypes

  @spec resolve(map(), map()) :: {:ok, %{url: String.t()}} | {:error, term()}
  def resolve(config, params) do
    AwsEndpointRules.evaluate(RuntimeTypes.endpoint_rule_set(), merge_params(config, params))
  end

  defp merge_params(config, params) do
    config_params = config_to_rule_params(config)
    client_params = client_context_params(config)
    Map.merge(Map.merge(config_params, client_params), params)
  end

  defp config_to_rule_params(config) do
    case Map.get(config, :region) do
      nil -> %{}

      value -> %{"Region" => value}
    end
  end

  defp client_context_params(config) do
    [optional_param(config, :region, "Region"), optional_param(config, :bucket, "Bucket")]
    |> Enum.reduce(%{}, fn map, acc -> Map.merge(acc, map) end)
  end

  defp optional_param(
    config,
    key,
    rule_key
  ) do
    case Map.get(config, key) do
      nil -> %{}

      value -> %{rule_key => value}
    end
  end
end